File: python/bench2-with-pipes.py
#!/usr/bin/env python3

import fileinput
import sys

set = []

def readDict():
  global set
  with open("../data/english_words.tsv") as f:
    for line in f:
      set.append(line.rstrip())
  set = frozenset(set)

def main():
  readDict()
  for line in sys.stdin:
    toks = line[:-1].split(' ') 
    for tok in toks:
      if tok in set:
        print ("match")

if __name__ == '__main__':
    main()

File: scala/bench2withpipes.scala
import java.io._
import scala.collection._

object bench2withpipes {

  val set = new mutable.HashSet[String]()

  def main(args:Array[String]) {
    readDict

    val reader = new BufferedReader(new InputStreamReader(System.in), 1024*1024)
    val writer = new BufferedWriter(new OutputStreamWriter(System.out), 1024*1024)

    var l:String = null
    while ({ l = reader.readLine; l != null }) {
      val toks = l.split(" ")
      for (tok <- toks)
        if (set.contains(tok))
          writer.write("match\n")
    }
    reader.close
    writer.close
  }

  def readDict = {
    val reader = new BufferedReader(new InputStreamReader(new FileInputStream("../data/english_words.tsv")))
    var l:String = null
    while ({ l = reader.readLine; l != null })
      set += l
    reader.close
  }

}
